src/internal/pachhash/fingerprinter.go

```diff
@@ -0,0 +1,49 @@
+package pachhash
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+)
+
+// KVFingerprinter is used to produce a unique collision-resistant fingerprint for a set of keys and values
+// The zero value of KVFingerprinter is ready to use.
+type KVFingerprinter struct {
+	sum  Output
+	last []byte
+}
+
+// Absorb adds a key, value pair to the fingerprinter, and sets the last key to key.
+// keys must be added in ascending order.
+func (kv *KVFingerprinter) Absorb(key, value []byte) error {
+	if kv.last != nil && bytes.Compare(kv.last, value) >= 0 {
```

Comment:
I feel like if last is nil you can still pass it to Compare, and it will be less than everything except nil.